### ISSUE TITLE
Unable to run horovod's keras example with the default docker image

### ISSUE THREAD

**Issue by suyash** [Azure/MachineLearningNotebooks#226]
I have already tried the distributed horovod example for TensorFlow checked in to this repository. It works for me.

Currently, I am trying to get the keras example from horovod to work (https://github.com/horovod/horovod/blob/master/examples/keras_mnist_advanced.py)

The default `azureml-defaults` doesn't work, because keras isn't included, so the run fails with

```
Traceback (most recent call last):
  File "azureml-setup/context_manager_injector.py", line 152, in <module>
    execute_with_context(cm_objects, options.invocation)
  File "azureml-setup/context_manager_injector.py", line 88, in execute_with_context
    runpy.run_path(sys.argv[0], globals(), run_name="__main__")
  File "/azureml-envs/azureml_ad47d8b4ce7b7db28fdf7352fc67bc29/lib/python3.6/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/azureml-envs/azureml_ad47d8b4ce7b7db28fdf7352fc67bc29/lib/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/azureml-envs/azureml_ad47d8b4ce7b7db28fdf7352fc67bc29/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "train.py", line 10, in <module>
    import horovod.keras as hvd
  File "/azureml-envs/azureml_ad47d8b4ce7b7db28fdf7352fc67bc29/lib/python3.6/site-packages/horovod/keras/__init__.py", line 16, in <module>
    import keras
ModuleNotFoundError: No module named 'keras'
```

Tried adding keras to both `conda_packages` or `pip_packages` in TensorFlow estimator, the build step is successful, however following that the training script errors out with

```
Traceback (most recent call last):
  File "azureml-setup/context_manager_injector.py", line 152, in <module>
    execute_with_context(cm_objects, options.invocation)
  File "azureml-setup/context_manager_injector.py", line 88, in execute_with_context
    runpy.run_path(sys.argv[0], globals(), run_name="__main__")
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "train.py", line 10, in <module>
    import horovod.keras as hvd
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/site-packages/horovod/keras/__init__.py", line 19, in <module>
    from horovod.tensorflow import init
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/site-packages/horovod/tensorflow/__init__.py", line 37, in <module>
    from horovod.tensorflow.mpi_ops import allgather, broadcast, _allreduce
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/site-packages/horovod/tensorflow/mpi_ops.py", line 58, in <module>
    ['HorovodAllgather', 'HorovodAllreduce'])
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/site-packages/horovod/tensorflow/mpi_ops.py", line 45, in _load_library
    library = load_library.load_op_library(filename)
  File "/azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/site-packages/tensorflow/python/framework/load_library.py", line 56, in load_op_library
    lib_handle = py_tf.TF_LoadLibrary(library_filename)
tensorflow.python.framework.errors_impl.NotFoundError: /azureml-envs/azureml_f616ff7896b421b3a238a203ef8c82cc/lib/python3.6/site-packages/horovod/tensorflow/mpi_lib.cpython-36m-x86_64-linux-gnu.so: undefined symbol: _ZNK10tensorflow8OpKernel4nameB5cxx11Ev
```

If I try to specify `horovod` in `pip_packages` or `conda_packages`, horovod installation in build step fails. A portion of it

```
...
INFO: Unable to build TensorFlow plugin, will skip it.
  
  Traceback (most recent call last):
    File "/tmp/pip-install-oukb9qgi/horovod/setup.py", line 55, in check_tf_version
      import tensorflow as tf
  ModuleNotFoundError: No module named 'tensorflow'
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/tmp/pip-install-oukb9qgi/horovod/setup.py", line 833, in build_extensions
      build_tf_extension(self, options)
    File "/tmp/pip-install-oukb9qgi/horovod/setup.py", line 566, in build_tf_extension
      check_tf_version()
    File "/tmp/pip-install-oukb9qgi/horovod/setup.py", line 62, in check_tf_version
      'import tensorflow failed, is it installed?\n\n%s' % traceback.format_exc())
  distutils.errors.DistutilsPlatformError: import tensorflow failed, is it installed?
  
  Traceback (most recent call last):
    File "/tmp/pip-install-oukb9qgi/horovod/setup.py", line 55, in check_tf_version
      import tensorflow as tf
  ModuleNotFoundError: No module named 'tensorflow'
...
```

Is there an example notebook/script working with keras?

**Comment by ninghu** [comment]
Hi @suyash,  you only need to add "keras" to the pip_packages if TensorFlow estimator is used. I just tried myself and it's working for me. Here is the code for your reference:

```
from azureml.train.dnn import TensorFlow

estimator= TensorFlow(source_directory=project_folder,
                      compute_target=compute_target,
                      entry_script='keras_mnist_advanced.py',
                      pip_packages=['keras'],
                      node_count=2,
                      process_count_per_node=1,
                      distributed_backend='mpi',
                      use_gpu=True)

```

**Comment by suyash** [comment]
@ninghu the difference between the code that I am trying to run and the code that is actually running is that I am using TensorFlow's keras implementation instead of the native keras implementation (`from tensorflow.keras.layers import Dense` instead of `from keras.layers import Dense`).

That's where the second problem I described happens. It seems that the keras implementation provided by the `keras` pip package is different and is giving me some errors in my own code, that I don't get with using the TensorFlow implementation. Closing this anyway, since horovod seems to be built to work with the `keras` pip package instead of the TensorFlow implementation.
